Refactor access to FEEvaluationImpl and friends

kronbichler opened the pull request:
As a preparation for #9794, I want to distinguish the calls to the various implementations with different `struct` names but the same `run()` function (to generically instantiate whatever is needed for the polynomial degrees). This PR makes the necessary refactoring to the cell and face evaluations, mostly just moving around some code.

~~I have set the `do not merge` label for now because I want to work on the downstream code tomorrow to see if there is anything else needed, but I wanted to announce it now to see if there are comments.~~

peterrum commented on include/deal.II/matrix_free/evaluation_kernels.h:
```diff
@@ -1967,27 +2291,71 @@ namespace internal
-            typename VectorizedArrayType,
-            typename Number2 = Number>
-  struct FEFaceEvaluationSelector
+  template <typename Number>
+  void
+  adjust_for_face_orientation(const unsigned int            dim,
```
Does it hurt if you make dim a template parameter?

kronbichler replied:
It's just a bit more code to be generated without any benefit - `dim` is only used for an outer loop.

peterrum reviewed (commented):
Looks good from my side. I think I know in which direction this is going!

Ping me once you have tried out the downstream code and added the promised tests! Looking forward to the next step!

kronbichler commented on include/deal.II/matrix_free/evaluation_kernels.h:
```diff
@@ -1377,6 +1377,330 @@ namespace internal
+        Number *gradients_quad,
+        Number *hessians_quad,
+        Number *scratch_data);
+  };
```
I am leaning towards using variadic template `args...` because the only way I see those could be generalized is by having a new struct that hides the function-specific part, which is also not pretty.

kronbichler commented:
I did two adjustments: I also applied the same interface to the inverse mass matrix and I changed the return type for all functions to `bool` because that is needed for the `gather_evaluate` call. It is not super pretty this way but we can tackle that in a later PR.

I consider this PR ready now.

> [...] and added the promised tests! 

That was a misleading comment from my side - I intended to say that I had to fix some of the failing existing tests (which is now concluded). The new interfaces are exercised by all code using `FEEvaluation` or `FEFaceEvaluation`, thus more than 400 tests (counting debug + release), which should be sufficient at this point. In a follow-up PR addressing #9794 I will switch a good bunch of those matrix-free tests to the non-templated version, so we stress even more parts.

peterrum reviewed (approved):
Looks good. As discussed offline, let's get #9794 working as soon as possible. We can clean up the internal data structures of `MatrixFree` afterwards. Let's try to distribute the task then fairly! Do you want to create a new issue so that we don't forget it. You can assign us both to it.

peterrum commented:
Thanks! One step closer to our goal!
